zilencer/views.py

```diff
@@ -1749,3 +1759,269 @@ def remote_server_check_analytics(request: HttpRequest, server: RemoteZulipServe
+    fcm_requests: list[firebase_messaging.Message],
+    fcm_remote_push_devices: list[RemotePushDevice],
+    delete_device_ids: list[int],
+) -> int:
+    try:
+        batch_response = firebase_messaging.send_each(fcm_requests, app=fcm_app)
+    except firebase_exceptions.FirebaseError:
+        logger.warning("Error while pushing to FCM", exc_info=True)
+        return 0
+
+    successfully_sent_count = 0
+    for idx, response in enumerate(batch_response.responses):
+        # We enumerate to have idx to track which token the response
+        # corresponds to. send_each() preserves the order of the messages,
+        # so this works.
+
+        remote_push_device = fcm_remote_push_devices[idx]
+        token = remote_push_device.token
+        push_account_id = remote_push_device.push_account_id
+        if response.success:
+            successfully_sent_count += 1
+            logger.info(
+                "FCM: Sent message with ID: %s to (push_account_id=%s, device=%s)",
+                response.message_id,
+                push_account_id,
+                token,
+            )
+        else:
+            error = response.exception
+            if isinstance(error, FCMUnregisteredError):
+                remote_push_device.expired_time = timezone_now()
+                remote_push_device.save(update_fields=["expired_time"])
+                delete_device_ids.append(remote_push_device.device_id)
+            else:
+                logger.warning(
+                    "FCM: Delivery failed for (push_account_id=%s, device=%s): %s:%s",
+                    push_account_id,
+                    token,
+                    error.__class__,
+                    error,
+                )
+
+    return successfully_sent_count
+
+
+def send_e2ee_push_notifications(
+    device_id_to_encrypted_data: dict[str, str],
+) -> SendNotificationResponseData:
+    import aioapns
+
+    device_ids = [int(device_id_str) for device_id_str in device_id_to_encrypted_data]
+    remote_push_devices = RemotePushDevice.objects.filter(
+        device_id__in=device_ids, expired_time__isnull=True
+    )
+    unexpired_remote_push_device_ids = [
+        remote_push_device.device_id for remote_push_device in remote_push_devices
+    ]
+
+    # Device IDs which should be deleted on server.
+    # Either the device ID is invalid or the token
+    # associated has been marked invalid/expired by APNs/FCM.
+    delete_device_ids = list(
+        filter(lambda device_id: device_id not in unexpired_remote_push_device_ids, device_ids)
+    )
+
+    apns_requests = []
+    apns_remote_push_devices: list[RemotePushDevice] = []
+    apns_base_message_payload = {
+        "aps": {
+            "mutable-content": 1,
+            "alert": {
+                "title": "New notification",
+            },
+            # TODO: Should we remove `sound` and let the clients add it.
+            # Then we can rename it as `apns_required_message_payload`.
+            "sound": "default",
```

Comment:
Probably a question for the chat.zulip.org thread.

Reply:
[#api design > E2EE - APNs payload @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/E2EE.20-.20APNs.20payload/near/2224183)